dotenv/src/iter.rs

```diff
@@ -38,39 +38,45 @@ struct QuotedLines<B> {
     buf: B,
 }
 
-enum QuoteState {
+enum ParseState {
     Complete,
     Escape,
     StrongOpen,
     StrongOpenEscape,
     WeakOpen,
     WeakOpenEscape,
+    Comment,
 }
 
-fn eval_end_state(prev_state: QuoteState, buf: &str) -> QuoteState {
+fn eval_end_state(prev_state: ParseState, buf: &str) -> ParseState {
     let mut cur_state = prev_state;
 
     for c in buf.chars() {
         cur_state = match cur_state {
-            QuoteState::Escape => QuoteState::Complete,
-            QuoteState::Complete => match c {
-                '\\' => QuoteState::Escape,
-                '"' => QuoteState::WeakOpen,
-                '\'' => QuoteState::StrongOpen,
-                _ => QuoteState::Complete,
+            ParseState::Escape => ParseState::Complete,
+            ParseState::Complete => match c {
+                '#' => return ParseState::Comment,
+                '\\' => ParseState::Escape,
+                '"' => ParseState::WeakOpen,
+                '\'' => ParseState::StrongOpen,
+                _ => ParseState::Complete,
             },
-            QuoteState::WeakOpen => match c {
-                '\\' => QuoteState::WeakOpenEscape,
-                '"' => QuoteState::Complete,
-                _ => QuoteState::WeakOpen,
+            ParseState::WeakOpen => match c {
+                '#' => return ParseState::Comment,
+                '\\' => ParseState::WeakOpenEscape,
+                '"' => ParseState::Complete,
+                _ => ParseState::WeakOpen,
             },
-            QuoteState::WeakOpenEscape => QuoteState::WeakOpen,
-            QuoteState::StrongOpen => match c {
-                '\\' => QuoteState::StrongOpenEscape,
-                '\'' => QuoteState::Complete,
-                _ => QuoteState::StrongOpen,
+            ParseState::WeakOpenEscape => ParseState::WeakOpen,
+            ParseState::StrongOpen => match c {
+                '#' => return ParseState::Comment,
```

Comment:
Here a regression is introduced: 
This line makes `KEY='ABC#DEF'` get truncated as `KEY='ABC`